### ISSUE TITLE
Anaconda prevents installlation from an ISO stored on same disk to a different partition (Install in unpartitioned space impossible)

### ISSUE THREAD

**Issue by tlaurion** [QubesOS/qubes-issues#8054]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### Qubes OS release
Q4.1, Q4.2

### Brief summary
Recent Heads permits to boot detached signed ISOs from local disk:
`media-scan /dev/sda1`

This works well to boot Tails, but anaconda refuses to install QubesOS on unpartitioned space.

### Expected behavior
Anaconda should not care about /dev/sda1 being mounted and permit to use unpartitioned space.

### Actual behavior
Anaconda prevents doing anything to /dev/sda since on same storage device

Will edit with screenshot and hypothesis/patches when testing new Q4.2 detached signed ISOs that have landed under https://qubes.notset.fr/iso/ today.

----

Starting with https://qubes.notset.fr/iso/Qubes-4.2.202302191645-x86_64.iso and https://qubes.notset.fr/iso/Qubes-4.2.202302191645-x86_64.iso.asc that should boot after being verified against https://github.com/osresearch/heads/blob/8b479b06ccd8afb3672f0f5d153fc27d44a12453/initrd/etc/distro/keys/qubes-testing.key

**Comment by tlaurion** [comment]
Q42 installer doesn't show behavior

**Comment by tlaurion** [comment]
@marmarek : as requested, issue opened. Unfortunately we just missed RC release :(

![signal-2023-02-20-152331](https://user-images.githubusercontent.com/827570/220202562-28966f26-7487-4f95-8a59-595a32b8035f.jpeg)
![signal-2023-02-20-152616](https://user-images.githubusercontent.com/827570/220202565-60eef8ee-c1d7-4fb4-bcd8-70e548c17e45.jpeg)

**Comment by tlaurion** [comment]
@marmarek: qubes 4.2 is bringing through anaconda the possibility to reclaim space from mounted partition.

Otherwise both 4.1 and 4.2 shows the bug: the installer is not using free space. Q4.2 automatically steals available space from /dev/sda1 but if no additional free space available from sda1, it will refuse to install.

**Comment by tlaurion** [comment]
@marmarek: no. I gave all space to /dev/sda1 and now the installer proposes me to reclaim space of sda1, but is unable because it is mounted as well.

Not sure I understand how it was happy with the 10 GB free inside of /dev/sda1 and reclaimed it automatically to install q4.2 but then failed on second stage install.

Trying same iso in USB installer, option 10 and will wipe install on /dev/sda

**Comment by tlaurion** [comment]
Heads booting from USB doesn't show behavior.

@marmarek : any insight on how to remove this limitation from anaconda?

**Comment by marmarek** [comment]
Was it working before? I'm pretty sure it didn't. Anyway, this can be implemented in R4.2, but backport to R4.1 depends on how complicated the change is.

**Comment by tlaurion** [comment]
@marmarek no it never worked before.


But having heads, not pre-provisioned, with a distinct partition offering both tails and latest qubesos directly available to boot from is really interesting for the whole "distrust everything" communities and would permit third party coreboot flashers to flash heads and distribute qubes and tails easily for DIY install/recovery/live situation for those who want Qubes available but not necessarily preinstalled.

Boot tails, download latest heads firmware. Verify qubes detached signature. Reboot. Reflash heads. Boot and install qubes. Factory reset USB security dongle. Enjoy Qubes.

**Comment by tlaurion** [comment]
@marmarek any direct pointers on where I would have to look to find source of the issue?

**Comment by marmarek** [comment]
Sorry, I don't know where exactly that's excluded. I'd start by looking at anaconda log (any of those in /tmp, possibly storage.log?) for messages about excluding that disk and then search in the code where it's done.
This might be related function: https://github.com/rhinstaller/anaconda/blob/master/pyanaconda/modules/storage/devicetree/model.py#L370 (see "hidden" and "protected" there).

**Comment by tlaurion** [comment]
Would have been handy again. Reinstalling from usb thumb drive, but left 30gb for /isos mount point this time so in can test this later on and provide logs for collaboration on testing this.


Heads validating iso integrity from usb takes way longer then from nvme, as expected. I guess deploying qubesos for OEMs woukd also be facilitated by this and Heads end users could reinstall from provided oem custom iso directly from that partition if bug was resolved.


Would also permit ends users to put detached signed isos under /isos partition, detached signed with their keys if distro don't provide such detached signatures, and reinstall blazingly fast.


And then maybe no need for oem re-ownership if heads could just boot from qubesos installer instead of reencrypting and reowning secrets? End users coukd just freshly install from verified qubesos image from fused qubes distro keys présent under heads. Same for booting tails if needed, which boots in amnesia state under 30 seconds in nv41 from nvme.

@marmarek

**Comment by tlaurion** [comment]
@marmarek issue lies in _mark_protected_devices marking whole drive being protected, not just the partition from which iso was read from.

Something like this would fix it (untested):
```

def _mark_protected_devices(self):
    """Mark protected devices.

    If a device is protected, mark it as such now. Once the tree
    has been populated, devices' protected attribute is how we will
    identify protected devices.
    """
    protected = []
    protected_with_ancestors = []

    # Resolve the protected device specs to devices.
    for spec in self.protected_devices:
        dev = self.devicetree.resolve_device(spec)

        if dev is not None:
            log.debug("Protected device spec %s resolved to %s.", spec, dev.name)
            protected.append(dev)

    # Find the stage2 backing device and its parents.
    stage2_device = find_stage2_device(self.devicetree)

    if stage2_device:
        log.debug("Resolved stage2 device to %s.", stage2_device.name)
        protected.append(stage2_device)

    # Find the live backing device and its parents.
    live_device = find_backing_device(self.devicetree, LIVE_MOUNT_POINT)

    if live_device:
        log.debug("Resolved live device to %s.", live_device.name)
        protected_with_ancestors.append(live_device)

    # Find the backing device of a stage2 source and its parents.
    source_device = find_backing_device(self.devicetree, DRACUT_REPO_DIR)

    if source_device:
        log.debug("Resolved a stage2 source device to %s.", source_device.name)
        protected.append(source_device)

    # For image installation setup_disk_images method marks all local
    # storage disks as ignored so they are protected from teardown.
    # Here we protect also cdrom devices from tearing down that, in case of
    # cdroms, involves unmounting which is undesirable (see bug #1671713).
    protected_with_ancestors.extend(dev for dev in self.devicetree.devices
                                    if dev.type == "cdrom")

    # Protect only the specific partition with an iso9660 file system. It will protect
    # NVDIMM devices that can be used only as an installation source anyway
    # (see the bug #1856264).
    for dev in self.devicetree.devices:
        if dev.format.type == "iso9660" and not dev.is_disk:
            log.debug("Marking partition %s with iso9660 as protected.", dev.name)
            protected_with_ancestors.append(dev)

    # Mark the collected devices as protected.
    for dev in protected:
        self._mark_protected_device(dev)
    for dev in protected_with_ancestors:
        self._mark_protected_device(dev, include_ancestors=True)
```

**Comment by ben-grande** [comment]
@tlaurion Affects R4.3?

**Comment by tlaurion** [comment]
> @tlaurion Affects R4.3?

Most probable @marmarek something changed? Doubt so.
